src/rnnlm/rnnlm-utils-test.cc

```diff
@@ -0,0 +1,177 @@
+// rnnlm/rnnlm-utils-test.cc
+
+#include <math.h>
+#include "rnnlm/rnnlm-utils.h"
+
+namespace kaldi {
+namespace rnnlm {
+
+void PrepareVector(int n, int ones_size, std::set<int> &outputs_set, vector<BaseFloat> &selection_probs) {
+  BaseFloat prob = 0;
+  BaseFloat prob_sum = 0;
+  for (int i = 0; i < n; i++) {
+    prob = RandUniform();
+    prob_sum += prob;
+    selection_probs.push_back(prob);
+  }
+  for (int i = 0; i < n; i++) {
+    selection_probs[i] /= prob_sum;
+  }
+  for (int i = 0; i < ones_size; i++) {
+    outputs_set.insert(rand() % n);
+  }
+}
+
+void UnitTestNChooseKSamplingConvergence(int n, int k, int ones_size) {
+  std::set<int> outputs_set;
+  vector<BaseFloat> selection_probs;
+  PrepareVector(n, ones_size, outputs_set, selection_probs);
+  NormalizeVec(k, outputs_set, &selection_probs);
+
+  vector<std::pair<int, BaseFloat> > u(selection_probs.size());
+  for (int i = 0; i < u.size(); i++) {
+    u[i].first = i;
+    u[i].second = selection_probs[i];
+  }
+  // normalize the selection_probs
+  BaseFloat sum = 0;
+  for (int i = 0; i < u.size(); i++) {
+    sum += selection_probs[i];
+  }
+  for (int i = 0; i < u.size(); i++) {
+    selection_probs[i] /= sum;
+  }
+
+  vector<BaseFloat> samples_probs(u.size(), 0);
+  int count = 0;
+  for (int i = 0; ; i++) {
+    count++;
+    vector<int> samples;
+    SampleWithoutReplacement(u, k, &samples);
+    for (int j = 0; j < samples.size(); j++) {
+      samples_probs[samples[j]] += 1;
+    }
+    for (int j = 0; j < samples_probs.size(); j++) {
```

Comment:
move this loop into the if(count % 1000 == 0) part - you're wasting computation here